### PR TITLE
[add-fire] Add learning rate and beta/epsilon decay to PyTorch

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

Adds linear decay parameters (if configured) to PPO and SAC. Note: we're not using the built in `torch.optim.lr_schedule` classes because those require the LR schedule be based on the number of epochs, and we want it to be based on number of steps. 

Compared to TF run (Blue is TF, Orange is Torch): 
image
image

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/torch/utils.py [hunk 1]
```diff
@@ -29,6 +29,66 @@ def swish(input_activation: torch.Tensor) -> torch.Tensor:
         """Swish activation function. For more info: https://arxiv.org/abs/1710.05941"""
         return torch.mul(input_activation, torch.sigmoid(input_activation))
 
+    @staticmethod
+    def apply_learning_rate(optim: torch.optim.Optimizer, lr: float) -> None:
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:33]: update_learning_rate ?

File: ml-agents/mlagents/trainers/torch/utils.py [hunk 2]
```diff
@@ -29,6 +29,66 @@ def swish(input_activation: torch.Tensor) -> torch.Tensor:
...
+        """
+        Apply a learning rate to a torch optimizer.
+        :param optim: Optimizer
+        :param lr: Learning rate
+        """
+        for param_group in optim.param_groups:
+            param_group["lr"] = lr
+
+    @staticmethod
+    def get_decayed_parameter(
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:43]: Would it make more sense to instantiate a DecayingParamter(torch.Parameter) that keeps the schedule, max_step, min_value, initial_value, etc... and that has an additional "update" method that takes the global_step (since it is the only thing that changes.

Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/utils.py:43]: I like this idea, since it would naturally save via state_dict. Currently it would have to go through the `update` at least once before the right `lr` is set.

Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/utils.py:43]: Hmm, on further thinking, there doesn't seem to be a real reason to make it an actual Parameter object though, since it won't be backpropp'ed through nor saved to an `nn.Module` (it's outside in the Optimizer). We can make it a DecayValue object with a `.get_value(global_step)`, and that should cut down the large function calls.

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 3]
```diff
@@ -65,7 +65,7 @@ def forward(
     def __init__(self, policy: TorchPolicy, trainer_params: TrainerSettings):
         super().__init__(policy, trainer_params)
         hyperparameters: SACSettings = cast(SACSettings, trainer_params.hyperparameters)
-        lr = hyperparameters.learning_rate
+        self.initial_lr = hyperparameters.learning_rate
         # lr_schedule = hyperparameters.learning_rate_schedule
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:69]:
```suggestion
```

Comment (andrewcoh) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:69]: Comments

File: ml-agents/mlagents/trainers/torch/utils.py (hunk 2 above)
Comment (dongruoping) [reply ml-agents/mlagents/trainers/torch/utils.py:43]: `optim.lr_scheduler`?

Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/utils.py:43]: +1

Comment (dongruoping) [reply ml-agents/mlagents/trainers/torch/utils.py:43]: `optim.lr_scheduler`?

File: ml-agents/mlagents/trainers/torch/utils.py [hunk 4]
```diff
@@ -29,6 +29,66 @@ def swish(input_activation: torch.Tensor) -> torch.Tensor:
...
+        else:
+            raise UnityTrainerException(f"The schedule {schedule} is invalid.")
+
+    @staticmethod
+    def polynomial_decay(
+        initial_value: float,
+        min_value: float,
+        max_step: int,
+        global_step: int,
+        power: float = 1.0,
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/torch/utils.py:75]: There's really no way to configure this properly. Should it even be an arg?

Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/utils.py:75]: If at some point we want a ScheduleType.EXPONENTIAL, we do

File: ml-agents/mlagents/trainers/torch/utils.py (hunk 2 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/utils.py:43]: I didn't go with `optim.lr_scheduler` since a) the API assumes lr is a function of the number of epochs, and we want to do it based on the number of steps, and b) `beta` and `epsilon` would still need another solution. Not really sure what the right answer is here. Just an `Object` seems strange but that seems to be all that is needed.
